pkg/events/derive/symbols_collision.go

```diff
@@ -0,0 +1,288 @@
+	if ok {
+		loadedObjects = loadedObjectsIface.([]sharedobjs.ObjInfo)
+		return loadedObjects, true
+	} else {
+		return loadedObjects, false
+	}
+}
+
+func (pcache *processesLoadedObjectsCache) SetProcessLoadedObjects(pid int, loadedObjects []sharedobjs.ObjInfo) {
+	pcache.cache.Add(pid, loadedObjects)
+}
+
+// sharedObjectsCollisionsCache is a cache for collision checks performed already
+// The results are saved per SO, which means that each collision is saved twice (because in each collision there are
+// 2 SOs which collide)
+type sharedObjectsCollisionsCache struct {
+	cache *simplelru.LRU
+}
+
+// AddCollisions add the collisions between 2 shared objects to the cache.
+func (socCache sharedObjectsCollisionsCache) AddCollisions(obj1 sharedobjs.ObjID,
+	obj2 sharedobjs.ObjID,
+	collisions []string) {
+	socCache.addObjCollisions(obj1, obj2, collisions)
+	socCache.addObjCollisions(obj2, obj1, collisions)
+}
+
+// GetCollision return the symbols collided between 2 shared objects from the cache if exist.
+// If the the collisions are cached for one object, will update it in the seconde object cache.
+// Return bool indication if the collision existed in the cache.
+func (socCache sharedObjectsCollisionsCache) GetCollision(obj1 sharedobjs.ObjID,
+	obj2 sharedobjs.ObjID) ([]string, bool) {
+	obj1Collisions, ok := socCache.getObjCollisions(obj1)
+	if ok {
+		collisions, ok := obj1Collisions[obj2]
+		if ok {
+			socCache.addObjCollisions(obj2, obj1, collisions)
+			return collisions, true
+		}
+	}
+	obj2Collisions, ok := socCache.getObjCollisions(obj2)
+	if ok {
+		collisions, ok := obj2Collisions[obj1]
+		if ok {
+			socCache.addObjCollisions(obj1, obj2, collisions)
+			return collisions, true
+		}
+	}
+	return nil, false
+}
+
+// Get from the cache the collisions check result of the object with other shared objects.
+// Return the collisions per other shared objects, and if the given shared object was monitored in the cache.
+func (socCache sharedObjectsCollisionsCache) getObjCollisions(objID sharedobjs.ObjID) (
+	map[sharedobjs.ObjID][]string, bool) {
+	var collisions map[sharedobjs.ObjID][]string
+	collisionsIface, ok := socCache.cache.Get(objID)
+	if ok {
+		collisions = collisionsIface.(map[sharedobjs.ObjID][]string)
+		return collisions, true
+	} else {
+		return collisions, false
+	}
+}
+
+func (socCache sharedObjectsCollisionsCache) setObjCollisions(objID sharedobjs.ObjID,
+	collisions map[sharedobjs.ObjID][]string) {
+	socCache.cache.Add(objID, collisions)
+}
+
+// addObjCollisions add to cache of one shared object collisions the collision with a second shared object.
+func (socCache sharedObjectsCollisionsCache) addObjCollisions(objID sharedobjs.ObjID,
+	collidedObjID sharedobjs.ObjID,
+	collisions []string) {
+	var objCollisions map[sharedobjs.ObjID][]string
+	collisionsIface, ok := socCache.cache.Get(objID)
+	if ok {
+		objCollisions = collisionsIface.(map[sharedobjs.ObjID][]string)
+		objCollisions[collidedObjID] = collisions
```

Comment:
can't this be removed as you're doing the same after if/else block ?

Reply:
I need to review my logic here, I think I have a bug.